File: frontend/src/api/objectives.ts
import { apiClient } from './client'

export type BusinessImpact = 'critical' | 'high' | 'medium' | 'low'

export interface ObjectiveEntry {
  type: string
  title: string
  description: string
  evidence_preview: string
  impact: string
  finding_ids: string[]
}

export interface EngagementObjectives {
  id: string
  achieved_objectives: ObjectiveEntry[]
  business_impact: BusinessImpact | null
  impact_narrative: string | null
  executive_summary: string | null
  remediation_plan: Array<{ finding_id: string; title: string; priority: string; effort: string }>
  operator_notes: string | null
  approved_by: string | null
  approved_at: string | null
  updated_at: string | null
}

export const objectivesApi = {
  get: (id: string) =>
    apiClient.get(`/api/engagements/${id}/phases/7/objectives`).then(r => r.data),

  save: (id: string, body: Partial<{
    achieved_objectives: ObjectiveEntry[]
    business_impact: BusinessImpact
    impact_narrative: string
    executive_summary: string
    remediation_plan: EngagementObjectives['remediation_plan']
    operator_notes: string
  }>) => apiClient.put(`/api/engagements/${id}/phases/7/objectives`, body).then(r => r.data),

  autoPopulate: (id: string) =>
    apiClient.post(`/api/engagements/${id}/phases/7/objectives/auto-populate`).then(r => r.data),

  signOff: (id: string, notes?: string) =>
    apiClient.post(`/api/engagements/${id}/phases/7/sign-off`, { notes }).then(r => r.data),
}

File: backend/app/models/objectives.py
import uuid
import enum
from datetime import datetime
from sqlalchemy import String, DateTime, Enum as SAEnum, ForeignKey, func, Text
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.dialects.postgresql import UUID, JSONB
from app.database import Base


class BusinessImpact(str, enum.Enum):
    critical = "critical"
    high     = "high"
    medium   = "medium"
    low      = "low"


class EngagementObjectives(Base):
    """Phase 7 — final impact documentation and executive summary."""
    __tablename__ = "engagement_objectives"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    engagement_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("engagements.id"), nullable=False, index=True, unique=True
    )
    phase_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("phases.id"), nullable=False
    )

    # List of {type, title, description, evidence_preview, impact, finding_ids}
    # type: data_exfil | privilege_escalation | rce | lateral_movement | persistence | credential_access | other
    achieved_objectives: Mapped[list] = mapped_column(JSONB, nullable=False, default=list)

    business_impact: Mapped[BusinessImpact | None] = mapped_column(
        SAEnum(BusinessImpact), nullable=True
    )

    # Free-form narrative for the technical audience
    impact_narrative: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Executive summary for non-technical stakeholders
    executive_summary: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Prioritised remediation list [{finding_id, title, priority, effort}]
    remediation_plan: Mapped[list] = mapped_column(JSONB, nullable=False, default=list)

    operator_notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    approved_by: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=True
    )
    approved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )
